pol-core/bscript/object.cpp

```diff
@@ -1653,6 +1655,99 @@ BObjectImp* ObjArray::call_method_id( const int id, Executor& ex, bool /*forcebu
         return new BError( "array.randomentry() doesn't take parameters." );
     }
     break;
+  case MTH_FILTER:
+    if ( name_arr.empty() )
+    {
+      if ( ex.numParams() < 1 )
+        return new BError( "Invalid parameter type" );
+
+      BObjectImp* param0 = ex.getParamImp( 0, BObjectType::OTFuncRef );
+
+      if ( !param0 )
+        return new BError( "Invalid parameter type" );
+
+      // The filter callback allows optional arguments, so no need to check the
+      // number of arguments for the passed function reference. Arguments passed
+      // will be shrunk and expanded (with uninit) as needed.
+
+      // If nothing to filter, return an empty array, since nothing to call the function with.
+      if ( ref_arr.empty() )
+        return new ObjArray();
+
+      // Arguments for user function call.
+      // - the element
+      // - the index of the element
+      // - the array itself
+      BObjectRefVec args;
+      args.push_back( ref_arr.front() );
+      args.push_back( BObjectRef( new BLong( 1 ) ) );
+      args.push_back( BObjectRef( this ) );
+
+      // The ContinuationCallback receives three arguments:
+      //
+      // - `Executor&`
+      // - `BContinuation* continuation`: The continuation, with methods to handle
+      //   the continuation (call the function again; finalize)
+      // - `BObjectRef result`: The result of the user function call specified in
+      //   `makeContinuation`.
+      //
+      // We pass to the lambda a reference to the element in case the user
+      // function modifies ref_arr.
+      //
+      // Returns a `BObjectImp`:
+      // - Call the user function again by returning the same continuation via
+      //   `ex.withContinuation`.
+      // - Return something else (in this case, the filtered array) to provide
+      //   that value back to the script.
+      auto callback = [this, filteredRef = BObjectRef( new ObjArray ),
+                       processedRef = BObjectRef( new BLong( 1 ) ),
```

Comment:
Since you create a new objectref for each run, so that the script cannot modify the index, wouldn't it be enough to just store/increment a plain integer?

Reply:
yes, but i couldn't figure out how to attach a "state member" to the lambda, eg:

```cpp
int processed = 0;

auto callback = [this, processed = std::move(processed)] () {
    processed++;
}
```
but this does not work

Reply:
> Since you create a new objectref for each run,

Also, is it really creating a new objectref for each run? I thought with eg. `processedRef = BObjectRef(...)` would only create this object when the lambda itself is created, and is "attached" to the lambda (and when the lambda gets freed, as do the bobjectrefs)